### PR TITLE
[Merged by Bors] - feat(RingTheory/LinearDisjoint): definition and properties of linearly disjoint of subalgebras (1)

### PR DESCRIPTION
**Pull request by acmepjz**
Main definitions:

- `Subalgebra.LinearDisjoint`: two subalgebras are linearly disjoint, if they are
  linearly disjoint as submodules (`Submodule.LinearDisjoint`).

- `Subalgebra.LinearDisjoint.mulMap`: if two subalgebras `A` and `B` of `S / R` are
  linearly disjoint, then there is `A ⊗[R] B ≃ₐ[R] A ⊔ B` induced by multiplication in `S`.

Equivalent characterization of linearly disjointness:

- `Subalgebra.LinearDisjoint.linearIndependent_left_of_flat`:
  if `A` and `B` are linearly disjoint, if `B` is a flat `R`-module, then for any family of
  `R`-linearly independent elements of `A`, they are also `B`-linearly independent.

- `Subalgebra.LinearDisjoint.of_basis_left`:
  conversely, if a basis of `A` is also `B`-linearly independent, then `A` and `B` are linearly disjoint.

- `Subalgebra.LinearDisjoint.linearIndependent_right_of_flat`:
  if `A` and `B` are linearly disjoint, if `A` is a flat `R`-module, then for any family of
  `R`-linearly independent elements of `B`, they are also `A`-linearly independent.

- `Subalgebra.LinearDisjoint.of_basis_right`:
  conversely, if a basis of `B` is also `A`-linearly independent,
  then `A` and `B` are linearly disjoint.

- `Subalgebra.LinearDisjoint.linearIndependent_mul_of_flat`:
  if `A` and `B` are linearly disjoint, if one of `A` and `B` is flat, then for any family of
  `R`-linearly independent elements `{ a_i }` of `A`, and any family of
  `R`-linearly independent elements `{ b_j }` of `B`, the family `{ a_i * b_j }` in `S` is
  also `R`-linearly independent.

- `Subalgebra.LinearDisjoint.of_basis_mul`:
  conversely, if `{ a_i }` is an `R`-basis of `A`, if `{ b_j }` is an `R`-basis of `B`,
  such that the family `{ a_i * b_j }` in `S` is `R`-linearly independent,
  then `A` and `B` are linearly disjoint.

Other main results:

- `Subalgebra.LinearDisjoint.symm_of_commute`, `Subalgebra.linearDisjoint_symm_of_commute`:
  linearly disjoint is symmetric under some commutative conditions.

- `Subalgebra.LinearDisjoint.bot_left`, `Subalgebra.LinearDisjoint.bot_right`:
  the image of `R` in `S` is linearly disjoint with any other subalgebras.

- `Subalgebra.LinearDisjoint.sup_free_of_free`: the compositum of two linearly disjoint
  subalgebras is a free module, if two subalgebras are also free modules.

- `Subalgebra.LinearDisjoint.inf_eq_bot_of_commute`, `Subalgebra.LinearDisjoint.inf_eq_bot`:
  if `A` and `B` are linearly disjoint, under suitable technical conditions, they are disjoint.

The results with name containing "of_commute" also have corresponding specified versions
assuming `S` is commutative.

---

discussion: <https://leanprover.zulipchat.com/#narrow/stream/217875-Is-there-code-for-X.3F/topic/Linearly.20disjoint>

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by erdOne** [comment]
Sorry for dropping the ball on this. This looks good to me. I'll wait if @kbuzzard wants to give it a look and I'll maintainer merge it this evening otherwise.

File: Mathlib/RingTheory/LinearDisjoint.lean [hunk 1]
```diff
@@ -0,0 +1,450 @@
...
+namespace LinearDisjoint
+
+variable [CommRing R] [Ring S] [Algebra R S]
+
+variable (A B : Subalgebra R S)
+
+lemma mulLeftMap_ker_eq_bot_iff_linearIndependent_op {ι : Type*} (a : ι → A) :
+    LinearMap.ker (Submodule.mulLeftMap (M := toSubmodule A) (toSubmodule B) a) = ⊥ ↔
+    LinearIndependent B.op (MulOpposite.op ∘ A.val ∘ a) := by
+  -- need this instance otherwise `LinearMap.ker_eq_bot` does not work
```
Comment (kbuzzard) [inline Mathlib/RingTheory/LinearDisjoint.lean:186]: The two letI lines can be removed if you start with
```lean
set_option maxSynthPendingDepth 2 in
lemma mulLeftMap_ker_eq_bot_iff_linearIndependent_op ...
```

File: Mathlib/RingTheory/LinearDisjoint.lean [hunk 2]
```diff
@@ -0,0 +1,450 @@
...
+  simp_rw [LinearIndependent, LinearMap.ker_eq_bot]
+  let i : (ι →₀ B) →ₗ[R] S := Submodule.mulLeftMap (M := toSubmodule A) (toSubmodule B) a
+  let j : (ι →₀ B) →ₗ[R] S := (MulOpposite.opLinearEquiv _).symm.toLinearMap ∘ₗ
+    (Finsupp.linearCombination B.op (MulOpposite.op ∘ A.val ∘ a)).restrictScalars R ∘ₗ
+    (Finsupp.mapRange.linearEquiv (linearEquivOp B)).toLinearMap
+  suffices i = j by
+    change Function.Injective i ↔ _
+    simp_rw [this, j, LinearMap.coe_comp,
+      LinearEquiv.coe_coe, EquivLike.comp_injective, EquivLike.injective_comp]
+    rfl
```
Comment (kbuzzard) [inline Mathlib/RingTheory/LinearDisjoint.lean:198]: I always feel like `rfl` after a simp call is abusing defeq. You can do
```suggestion
    simp_rw [this, j, LinearMap.coe_comp,
      LinearEquiv.coe_coe, EquivLike.comp_injective, EquivLike.injective_comp,
      LinearMap.coe_restrictScalars]
```
instead.

File: Mathlib/RingTheory/LinearDisjoint.lean [hunk 3]
```diff
@@ -0,0 +1,450 @@
...
+theorem of_basis_left_op {ι : Type*} (a : Basis ι R A)
+    (H : LinearIndependent B.op (MulOpposite.op ∘ A.val ∘ a)) :
+    A.LinearDisjoint B := by
+  rw [← mulLeftMap_ker_eq_bot_iff_linearIndependent_op] at H
+  exact Submodule.LinearDisjoint.of_basis_left _ _ a H
+
+lemma mulRightMap_ker_eq_bot_iff_linearIndependent {ι : Type*} (b : ι → B) :
+    LinearMap.ker (Submodule.mulRightMap (toSubmodule A) (N := toSubmodule B) b) = ⊥ ↔
+    LinearIndependent A (B.val ∘ b) := by
+  -- need this instance otherwise `LinearMap.ker_eq_bot` does not work
```
Comment (kbuzzard) [inline Mathlib/RingTheory/LinearDisjoint.lean:228]: The `set_option maxSynthPendingDepth 2 in` trick also works here.

**Review by kbuzzard — APPROVED** [review]
This is a great piece of work!

File: Mathlib/RingTheory/LinearDisjoint.lean (hunk 1 above)
Comment (acmepjz) [reply Mathlib/RingTheory/LinearDisjoint.lean:186]: Thank you!

**Comment by acmepjz** [comment]
Done.

**Comment by acmepjz** [comment]
**ping** what's the progress?

**Comment by erdOne** [comment]
Thanks!
maintainer merge

File: Mathlib/RingTheory/LinearDisjoint.lean [hunk 4]
```diff
@@ -0,0 +1,448 @@
...
+
+- `Subalgebra.LinearDisjoint.mulMap`: if two subalgebras `A` and `B` of `S / R` are
+  linearly disjoint, then there is `A ⊗[R] B ≃ₐ[R] A ⊔ B` induced by multiplication in `S`.
+
+## Main results
+
+### Equivalent characterization of linearly disjointness
+
+- `Subalgebra.LinearDisjoint.linearIndependent_left_of_flat`:
+  if `A` and `B` are linearly disjoint, if `B` is a flat `R`-module, then for any family of
```
Comment (jcommelin) [inline Mathlib/RingTheory/LinearDisjoint.lean:36]:
```suggestion
  if `A` and `B` are linearly disjoint, and if `B` is a flat `R`-module, then for any family of
```
I hope this doesn't make the :100: chars linter angry.

File: Mathlib/RingTheory/LinearDisjoint.lean [hunk 5]
```diff
@@ -0,0 +1,448 @@
...
+  `R`-linearly independent elements of `A`, they are also `B`-linearly independent.
+
+- `Subalgebra.LinearDisjoint.of_basis_left_op`:
+  conversely, if a basis of `A` is also `B`-linearly independent, then `A` and `B` are
+  linearly disjoint.
+
+- `Subalgebra.LinearDisjoint.linearIndependent_right_of_flat`:
+  if `A` and `B` are linearly disjoint, if `A` is a flat `R`-module, then for any family of
```
Comment (jcommelin) [inline Mathlib/RingTheory/LinearDisjoint.lean:44]:
```suggestion
  if `A` and `B` are linearly disjoint, and if `A` is a flat `R`-module, then for any family of
```
I hope this doesn't make the :100: chars linter angry.

File: Mathlib/RingTheory/LinearDisjoint.lean [hunk 6]
```diff
@@ -0,0 +1,448 @@
...
+  `R`-linearly independent elements of `B`, they are also `A`-linearly independent.
+
+- `Subalgebra.LinearDisjoint.of_basis_right`:
+  conversely, if a basis of `B` is also `A`-linearly independent,
+  then `A` and `B` are linearly disjoint.
+
+- `Subalgebra.LinearDisjoint.linearIndependent_mul_of_flat`:
+  if `A` and `B` are linearly disjoint, if one of `A` and `B` is flat, then for any family of
```
Comment (jcommelin) [inline Mathlib/RingTheory/LinearDisjoint.lean:52]:
```suggestion
  if `A` and `B` are linearly disjoint, and if one of `A` and `B` is flat, then for any family of
```
I hope this doesn't make the :100: chars linter angry.

**Review by jcommelin — COMMENTED** [review]
bors d+

**Pull request closed without merging**